From 98825110cedf6f0de4b8245cab68b5aea5e1b6b3 Mon Sep 17 00:00:00 2001
From: Jiri Olsa <jolsa@redhat.com>
Date: Tue, 19 Aug 2014 15:25:54 -0400
Subject: [tools] perf/annotate: Find kcore symbols on other maps

Message-id: <1408462094-14194-186-git-send-email-jolsa@redhat.com>
Patchwork-id: 88205
O-Subject: [PATCH RHEL7.1 BZ1131394 185/325] perf annotate: Find kcore symbols on other maps
Bugzilla: 1131394
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Jiri Olsa <jolsa@kernel.org>

Bugzilla: 1131394
https://bugzilla.redhat.com/show_bug.cgi?id=1131394

upstream
========
commit 6e427ab02c8886ca6c9ecdbb318e68fe8f605469
Author: Adrian Hunter <adrian.hunter@intel.com>
Date: Mon Oct 14 13:43:40 2013 +0300

description
===========
Use the new map_groups__find_ams() method to find kcore symbols on other
maps.
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/tools/perf/util/annotate.c b/tools/perf/util/annotate.c
index 882bb86..cf6242c 100644
--- a/tools/perf/util/annotate.c
+++ b/tools/perf/util/annotate.c
@@ -825,20 +825,16 @@ static int symbol__parse_objdump_line(struct symbol *sym, struct map *map,
   dl->ops.target.offset = dl->ops.target.addr -
      map__rip_2objdump(map, sym->start);
 
- /*
-  * kcore has no symbols, so add the call target name if it is on the
-  * same map.
-  */
+ /* kcore has no symbols, so add the call target name */
  if (dl->ins && ins__is_call(dl->ins) && !dl->ops.target.name) {
-  struct symbol *s;
-  u64 ip = dl->ops.target.addr;
-
-  if (ip >= map->start && ip <= map->end) {
-   ip = map->map_ip(map, ip);
-   s = map__find_symbol(map, ip, NULL);
-   if (s && s->start == ip)
-    dl->ops.target.name = strdup(s->name);
-  }
+  struct addr_map_symbol target = {
+   .map = map,
+   .addr = dl->ops.target.addr,
+  };
+
+  if (!map_groups__find_ams(&target, NULL) &&
+      target.sym->start == target.al_addr)
+   dl->ops.target.name = strdup(target.sym->name);
  }
 
  disasm__add(&notes->src->source, dl);
-- 
1.7.1